metrics_batcher.go

```diff
@@ -25,55 +25,90 @@ import (
 
 type metricsBatcher struct {
 	projectName string
-	allReqs     []*monitoringpb.CreateTimeSeriesRequest
 	allTss      []*monitoringpb.TimeSeries
 	allErrs     []error
-	// Counts all dropped TimeSeries by this exporter.
+
+	// Counts all dropped TimeSeries by this metricsBatcher.
 	droppedTimeSeries int
+
+	workers   []*worker
+	curWorker int
+
+	mc *monitoring.MetricClient
 }
 
-func newMetricsBatcher(projectID string) *metricsBatcher {
+func newMetricsBatcher(ctx context.Context, projectID string, numWorkers int, mc *monitoring.MetricClient) *metricsBatcher {
+	workers := make([]*worker, 0, numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		w := newWorker(ctx, mc)
+		workers = append(workers, w)
+		go w.start()
+	}
 	return &metricsBatcher{
 		projectName:       fmt.Sprintf("projects/%s", projectID),
 		allTss:            make([]*monitoringpb.TimeSeries, 0, maxTimeSeriesPerUpload),
 		droppedTimeSeries: 0,
+		workers:           workers,
+		curWorker:         0,
+		mc:                mc,
 	}
 }
 
-func (mb *metricsBatcher) recordDroppedTimeseries(numTimeSeries int, err error) {
+func (mb *metricsBatcher) recordDroppedTimeseries(numTimeSeries int, errs ...error) {
 	mb.droppedTimeSeries += numTimeSeries
-	mb.allErrs = append(mb.allErrs, err)
+	for _, err := range errs {
+		if err != nil {
+			mb.allErrs = append(mb.allErrs, err)
+		}
+	}
 }
 
 func (mb *metricsBatcher) addTimeSeries(ts *monitoringpb.TimeSeries) {
 	mb.allTss = append(mb.allTss, ts)
-	if len(mb.allTss) == maxTimeSeriesPerUpload {
-		mb.allReqs = append(mb.allReqs, &monitoringpb.CreateTimeSeriesRequest{
+	if len(mb.allTss) == maxTimeSeriesPerUpload && len(mb.workers) != 0 {
+		req := &monitoringpb.CreateTimeSeriesRequest{
 			Name:       mb.projectName,
 			TimeSeries: mb.allTss,
-		})
+		}
+		mb.workers[mb.curWorker].reqChan <- req
+		mb.nextWorker()
 		mb.allTss = make([]*monitoringpb.TimeSeries, 0, maxTimeSeriesPerUpload)
 	}
 }
 
-func (mb *metricsBatcher) export(ctx context.Context, mc *monitoring.MetricClient) {
-	// Last batch, if any.
-	if len(mb.allTss) > 0 {
-		mb.allReqs = append(mb.allReqs, &monitoringpb.CreateTimeSeriesRequest{
-			Name:       mb.projectName,
-			TimeSeries: mb.allTss,
-		})
+func (mb *metricsBatcher) nextWorker() {
+	mb.curWorker++
+	if mb.curWorker >= len(mb.workers) {
+		mb.curWorker = 0
 	}
+}
 
-	// Send create time series requests to Stackdriver.
-	for _, req := range mb.allReqs {
-		if err := createTimeSeries(ctx, mc, req); err != nil {
-			mb.recordDroppedTimeseries(len(req.TimeSeries), err)
+func (mb *metricsBatcher) close(ctx context.Context) error {
+	for _, w := range mb.workers {
+		resp := w.stop()
+		mb.recordDroppedTimeseries(resp.droppedTimeSeries, resp.errs...)
+	}
+
+	// Send any remaining time series
+	if len(mb.allTss) > 0 && mb.mc != nil {
```

Comment:
Why not adding it to the workers?

Reply:
There're two cases here:
1) If there're any workers, the last batch must have < 200 timeseries, which can only be converted to one request. However even if we pass the requests to workers, this function `close` has to wait for all workers to finish in order to count the errors and dropped timeseries. There's not much benefit to delegate it to workers given the cost of putting it onto channel then block waiting for response, as far as I can see.
2) If there's no worker, the last batch could have >=200 timeseries and be converted to >1 requests. But since there's no worker this has to be done in this goroutine.

Reply:
num_workers should always be >=1 so the number 2 does not apply (or at least in what I thought the implementation would look like).

For 1. I agree that you do wait in the end for all but with my proposal:
* Do not enforce that all previous requests are finished;
* No special logic for this request.

Reply:
Indeed, I was wrong, let's just simplify and always have at least 1 worker.

Reply:
Updated in https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/pull/213/commits/8a3628b2c98c2730757638fededf6eab96624ece.